Add type hints (mypy assumed as reference/standard)

abravalheri opened the pull request:
This is an attempt to start addressing issue #16.

Adding type checking to an untyped codebase is usually a challenge and
sometimes require pragmatic approaches, such as the one adopted in this
commit.

`# type: ignore` and `cast` are used sporadically, but still present when the
type checker is not "smart enough", when the current Python syntax does
not allow expressing the most correct type information or when it was
too hard to introduce it.

It was necessary to remove explicit inheritance from `MuttableMapping`,
since the abstract methods are being overwritten with a non matching
signature that is more convenient for ConfigUpdater.

Functions that use `_UNSET` need to be annotated using `overload`

---
As a quick report, the process used to produce the type annotations was
"semi-manual", with the following steps:

1. Using `pytype` as type checker, the type errors were debugged and
   fixed or "workaround"-ed
2. Using `pytype`, initial type stubs were created
3. Using `retype`, the type stubs were merged back into the source code
   (for some reason `pytype`'s `merge-pyi` did not work)
4. Manual checking/refactoring/fixing of the annotated file
5. Introduction of new annotations not generated by `pytype`
6. Using `mypy` as type checker, the type errors were debugged and fixed
   or "workaround"-ed

mypy does seem to have a tool for generating type stubs, but I have
never worked with it...

abravalheri commented:
So... there is a problem with removing `MutableMapping`: although in a quick look at the docs the class just seem to be an interface, it is actually providing some methods like `pop` which is being used by PyScaffold.

However, if we just make `Section <: MutableMapping[str, Option]` then we have loads of typecheck errors
because the following is expected:
```python
def __setvalue__(self, key: str, value: Option): ...
def __iter__(self) -> Iterator[str]: ...
```

and the current implementation is:
```python
def __setvalue__(self, key: str, value: Optional[str]): ...
def __iter__(self) -> Iterator[SectionContents]: ...
```

For `ConfigUpdater <: MutableMapping[str, Section]` it is even worse, because the `get` method is completely being taken over for a different task:

expected:
```python
def __iter__(self) -> Iterator[str]: ...
def get(self, key: str, default: T) -> Union[Section, T]: ...
```

current implementation:
```python
def __iter__(self) -> Iterator[ConfigContents]: ...
def get(self, section: str, option: str, fallback: T) -> Union[Option, T]: ...
```

Unfortunately `ConfigParser` seems to violate the Liskov substition principle, and `ConfigUpdater` follows the same path 😢

abravalheri commented:
If we are willing to change the API, my suggestion would be keep the signatures as defined in `MutableMapping` and define other methods with other names for the intended functionality, for example:

```python
class Container(Generic[T]):
    ...
    def iter_blocks(self) -> T:
        return iter(self._structure)
    ...

class Section(Block[ConfigContent], Container[SectionContent], MutableMapping[str, Option]):
    ...
    def set(self: T, key: str, value: str) -> T:
    ...

class ConfigUpdater(Container[ConfigContent], MutableMapping[str, Section]):
    ...
    def get_option(self: T, section: str, option: str, fallback: T) -> Union[Option, T]:
    ...
```
and so on for all the methods not following the Liskov substition principle...

However, this also means reducing the compatibility with `ConfigParser` (previously it was almost a drop-in replacement for `ConfigParser`)...

abravalheri commented:
`ConfigParser` seems to simply [ignore the problem](https://github.com/python/typeshed/blob/faa5c70f25986e11c59e573dfcf17e3052393c76/stdlib/configparser.pyi#L140).

abravalheri commented on src/configupdater/configupdater.py:
```diff
@@ -30,8 +30,7 @@
 from abc import ABC
-from collections import OrderedDict as _default_dict
-from collections.abc import MutableMapping
+from collections import OrderedDict
```
From Python 3.6 onwards `dict` is ordered by default... is `OrderedDict` still necessary?

FlorianWilhelm replied:
I would say so as it is more explicit. I read somewhere that no one guarantees that this behaviour will not change in future Python versions.

FlorianWilhelm commented:
Thanks, @abravalheri for putting so much effort into this. My take would be to stay as close as possible to ConfigParser and rather follow a design flaw by design flaw kind of approach, so if ConfigParser violates something, we do it too in order to follow the Principle of Least Astonishment. Does this seem reasonable to you?

FlorianWilhelm commented on src/configupdater/configupdater.py:
```diff
@@ -289,16 +323,17 @@ def __str__(self):
+        try:
+            return next(o for o in self.iter_options() if o.key == key)
+        except StopIteration:
+            raise KeyError("No option `{}` found".format(key), {"key": key})
```
Would it make sense to do
```
except StopIteration as e:
  raise KeyError("No option `{}` found".format(key), {"key": key}) from e
```
Using raise from helps when debugging as it keeps the stack trace.

FlorianWilhelm commented on src/configupdater/configupdater.py:
```diff
@@ -307,59 +342,74 @@ def __setitem__(self, key, value):
+            idx = self._get_option_idx(key=key)
+            del self._structure[idx]
+        except StopIteration:
+            raise KeyError("No option `{}` found".format(key), {"key": key})
```
raise from?

abravalheri commented:
I suppose we can do that 😝 
(This would solve the overwriten `get`, I would still go for separated methods for `__iter__`).
 
I will work a bit more on this PR and follow shortly with a more complete approach.

---
Just for the sake of brainstrom, we could also do:

Suppose we have a hashable `HiddenKey` class (dataclasses can be used for that), we can apply polymorphism to `Key = Union[HiddenKey, str]` and always associate `Comment` and `Space` with an auto-generated `HiddenKey`... this way we could make
`Section <: MutableMapping[Key, SectionContents]` and `ConfigUpdater <: MutableMapping[Key, ConfigContents]`.

(I guess using this we can even simplify/speed-up lookups and avoid constantly looping, just relying on standard dict hashing).

I will not work on that, but please feel free to share your thoughts.

FlorianWilhelm commented:
Regarding `__iter__` I think it should be consistent with accessing an item and there it makes sense to return a `Block` as people are using ConfigUpdater to change something. How about one no longer inherits from `MutableMapping` and reimplements the few methods it provides currently as a MixinClass? Wouldn't this solve it?

Regarding the brainstorming idea. I am always in favor of keeping things simple. I think I don't understand it. Associating `Comment` and `Space` with auto-generated `HiddenKey` and then using Polymorphism... Isn't that the same as saying `Key = Union[Block, str]` since `Comment` and `Space` inherit from `Block` and thus are Blocks? Okay, then maybe the type checker would complain about calling methods that are not defined in `Block`. Couldn't we just list all blocks, e.g. `Key = Union[Comment, Space, Section, Option, str]`

abravalheri commented:
Ok, so this last commit was exploring the previous idea. I am defining a separated `iter_blocks` and leaving `__iter__` for `MutableMapping`. I still haven't had the time to address/comment your previous comment but I will go back to that later today.

I tested the implementation of https://github.com/pyscaffold/configupdater/pull/23/commits/bbac2def1c65592bc1063e233d658fb238d35b73 against PyScaffold's test suite and it does work... (we are not relying on `__iter__`).

I also added a bunch of methods for a more familiar (`dict`-like) `get` experience.

FlorianWilhelm commented:
Thanks @abravalheri. Let's discuss another day. My point only is that `__getitem__` and `__iter__` should be consistent like you have it in lists, arrays, and all kinds of containers accessed by some key or index.

abravalheri commented:
No problems @FlorianWilhelm, we can even have a call one of these days. I will try to quickly summarise my ideas below (regarding the type inconsistencies):

In my mind the problem that we are facing is related to combining two different concepts/protocol/interfaces that define the same methods: both `Section` and `ConfigUpdater` try to be a `Sequence[Block]` and a `Mapping[str, T]` (with `T = Option` for `Sequence` and `T = Section` for `ConfigUpdater`), and unfortunately the `Sequence` and `Mapping` protocols take over the `__iter__` method, so this combination cannot coexist...

We can argue that the code does no enforce `Sequence[Block]`, but in the end of the day, that is what the `__iter__` is currently implementing. I suppose `ConfigParser` does not have this problem because it simply ignores comments and spacing and therefore can drop the `Sequence` part and stick with `Mapping`.

With that in mind, I would actually advocate for keeping `__iter__(self) -> Iterable[str]`, because that is more compatible with `ConfigParser` **and** doing otherwise would imply that the current implementations of `__getitem__`, `__delitem__`, `items`, etc are not coherent with `__iter__` (you would be able to iterate over items you cannot get or delete, for example).
That is the approach I followed in the commits so far, and to compensate I have introduced `iter_blocks`.

Regarding `get`, I followed you advice and simply silenced the type checker complaints, since that is what `ConfigParser` does.
Since in the past while invoking this method from PyScaffold I was really frustrated with it completely ignoring the Principle of Least Astonishment and offering an API that looks counter-intuitive for someone used to Python, I added some methods that would give me back the `dict` behaviour 😝 : `ConfigUpdater.get_section` is what `ConfigUpdater.get` should be... (in another world the existing method should be named `get_option` for clarity).

Now, if we are really concerned with `ConfigParser` compatibility, ideally `ConfigUpdate.get` and `Section.get` should return `str` instead of `Option`, otherwise the users will be surprised and always forget to access `.value` ... I added a `Section.get_option(key: str, default: T) -> Union[str, T, None]` to compensate for that, but still it is something to be discussed.

abravalheri commented:
Regarding the brainstorm, what I was thinking could be translated to the following pseudo-code for AST:
```python
config = ConfigUpdater(_structure={
     HiddenKey(00): Comment('# This file ....'),
     "metadata": Section(name="metadata", _structure={
         "name": Option(value="configupdater", ...),
         HiddenKey(10): Comment(...),
         HiddenKey(11): Space(...),
    }),
    HiddenKey(01): Space(...),
    "options": Section(name="options", _structure={...}),
    HiddenKey(02): Space(...),
    HiddenKey(03): Comment(...),
    ....
})
# i.e. ConfigUpdater <: MutableMapping[Union[str, HiddenKey], Union[Section, Comment, Space]]
# this way Comments and Spaces are comparable to Section/Options
```

If I understood correctly, what you described would be:
```python
config = ConfigUpdater(_structure={
     Comment('# This file ....'): None,
     "metadata": Section(name="metadata", _structure={
         "name": Option(value="configupdater", ...),
         Comment(...): None,
         Space(...): None,
    }),
    Space(...): None,
    "options": Section(name="options", _structure={...}),
    Space(...): None,
    Comment(...): None,
    ....
})
# i.e. ConfigUpdater <: MutableMapping[Union[str, Comment, Space], Optional[Section]]
# this way Comments and Spaces are comparable to section and option names/keys
```

abravalheri commented:
Sorry more commits and more text explaining :P

The last changes try to revert the implementation to the same API in v2.0. I added the original behaviour to __iter__ back and removed the `get_option` method I have previously added because the name was too confusing for a public API.

I also added in the docstring some indication to the fact the API is not following Liskov substition principle.
Hopefully the implementation as it is in this branch should be completely compatible with the API in v2.0.

FlorianWilhelm commented:
Thanks, @abravalheri for all your work and the amazing effort you put into this. I was work-wise quite busy the last few days and I would like to take you up on your idea to maybe have another video call to discuss the changes. 

You are definitely right about the `__iter__` and `__getitem__` consistency, I got it mixed up and I am sorry for that. I think I need some more time to really wrap my head around your suggestions and then let's have a video call maybe in about a week or so to discuss it. Have a great weekend until then :-)
